core/src/main/java/com/datastax/oss/driver/internal/core/metadata/DefaultTopologyMonitor.java

```diff
@@ -259,80 +256,124 @@ private String retrievePeerTableName() {
     return "system.peers";
   }
 
-  private NodeInfo asNodeInfo(AdminRow row) {
-    return nodeInfoBuilder(row, getBroadcastRpcAddress(row)).build();
-  }
-
-  private Optional<NodeInfo> firstRowAsNodeInfo(AdminResult result) {
+  private Optional<NodeInfo> firstPeerRowAsNodeInfo(AdminResult result) {
     Iterator<AdminRow> iterator = result.iterator();
     if (iterator.hasNext()) {
-      return Optional.of(asNodeInfo(iterator.next()));
-    } else {
-      return Optional.empty();
+      AdminRow row = iterator.next();
+      if (isPeerValid(row)) {
+        InetSocketAddress peerBroadcastRpcAddress = getBroadcastRpcAddress(row);
+        return Optional.of(nodeInfoBuilder(row, peerBroadcastRpcAddress).build());
+      }
     }
+    return Optional.empty();
   }
 
   /**
+   * Creates a {@link DefaultNodeInfo.Builder} instance from the given row.
+   *
    * @param broadcastRpcAddress this is a parameter only because we already have it when we come
    *     from {@link #findInPeers(AdminResult, InetSocketAddress)}. Callers that don't already have
-   *     it can use {@link #getBroadcastRpcAddress}.
+   *     it can use {@link #getBroadcastRpcAddress}. For the control host, this can be null; if this
+   *     node is a peer however, this cannot be null, since we use that address to create the node's
+   *     endpoint. Callers can use {@link #isPeerValid(AdminRow)} to check that before calling this
+   *     method.
    */
+  @NonNull
   protected DefaultNodeInfo.Builder nodeInfoBuilder(
-      AdminRow row, InetSocketAddress broadcastRpcAddress) {
-
-    // Deployments that use a custom EndPoint implementation will need their own TopologyMonitor.
-    // One simple approach is to extend this class and override this method.
-    EndPoint endPoint =
-        new DefaultEndPoint(context.getAddressTranslator().translate(broadcastRpcAddress));
-
-    DefaultNodeInfo.Builder builder =
-        DefaultNodeInfo.builder()
-            .withEndPoint(endPoint)
-            .withBroadcastRpcAddress(broadcastRpcAddress);
-    InetAddress broadcastAddress = row.getInetAddress("broadcast_address"); // in system.local
-    if (broadcastAddress == null) {
-      broadcastAddress = row.getInetAddress("peer"); // in system.peers
+      @NonNull AdminRow row, @Nullable InetSocketAddress broadcastRpcAddress) {
+
+    boolean peer = row.contains("peer");
+
+    EndPoint endPoint;
+    if (peer) {
+      // If this node is a peer, its broadcast RPC address must be present
+      assert broadcastRpcAddress != null;
```

Comment:
At the line you mention `broadcastRpcAddress` cannot be null since it was tested by `isValidPeer`. All calls to `nodeInfoBuilder` are protected by `isValidPeer` except [this one](https://github.com/datastax/java-driver/blob/cd295a34979e122a94da6d04aa32c1067d545e79/core/src/main/java/com/datastax/oss/driver/internal/core/metadata/DefaultTopologyMonitor.java#L201) because this one is the local (control) host.